File: scripts/hooks_doctor.py
#!/usr/bin/env python3
"""Hook doctor — read-only diagnostic over the hook runtime.

Wraps `scripts/hooks_status.py` (bridge presence + manifest bindings)
and adds three diagnostics the bare status table does not surface:

  * **Concerns** — every concern declared in the manifest, its
    `fail_closed` posture, the on-disk script path, and a one-line
    file-exists check.
  * **Trampolines** — per-platform shell trampoline expected under
    `scripts/hooks/<platform>-dispatcher.sh`; flags any platform that
    has manifest bindings but no trampoline on disk.
  * **Last feedback** — for each concern, the most-recent dispatcher
    feedback file under `agents/runtime/state/.dispatcher/*/<concern>.json`,
    plus the per-rule state file under `agents/runtime/state/<concern>.json`
    when one exists.

This is a **read-only** report. It never installs, modifies, or runs
anything — same contract as `hooks_status.py`. CI uses `--strict` to
turn missing bindings / trampolines into a non-zero exit.

Schema: docs/contracts/hook-architecture-v1.md.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPO_ROOT / "scripts"))
sys.path.insert(0, str(REPO_ROOT / "scripts" / "hooks"))

import dispatch_hook  # noqa: E402
import hooks_status  # noqa: E402

TRAMPOLINE_DIR = REPO_ROOT / "scripts" / "hooks"
STATE_DIR_DEFAULT = "agents/runtime/state"

# Platforms whose bridge file (settings.json) invokes the universal
# dispatcher directly — no shell trampoline required. Excluded from the
# "missing trampoline" check.
NATIVE_DISPATCH_PLATFORMS = frozenset({"claude"})


def _trampoline_for(platform: str) -> Path:
    return TRAMPOLINE_DIR / f"{platform}-dispatcher.sh"


def _concern_state_file(state_dir: Path, concern: str) -> Path | None:
    target = state_dir / f"{concern}.json"
    return target if target.is_file() else None


def _latest_feedback(state_dir: Path, concern: str) -> Path | None:
    """Return the most-recent dispatcher feedback file for the concern,
    walking `agents/runtime/state/.dispatcher/<session>/<concern>.json`."""
    dispatcher_dir = state_dir / ".dispatcher"
    if not dispatcher_dir.is_dir():
        return None
    candidates = sorted(
        dispatcher_dir.glob(f"*/{concern}.json"),
        key=lambda p: p.stat().st_mtime,
        reverse=True,
    )
    return candidates[0] if candidates else None


def _rel(path: Path | None, root: Path) -> str | None:
    if path is None:
        return None
    try:
        return str(path.resolve().relative_to(root.resolve()))
    except ValueError:
        return str(path)


def collect(project_root: Path, manifest: dict,
            state_dir_rel: str = STATE_DIR_DEFAULT) -> dict:
    """Build the doctor payload — JSON-serialisable."""
    matrix = hooks_status.collect(project_root, manifest)
    state_dir = project_root / state_dir_rel

    concerns_def = manifest.get("concerns") or {}
    concerns: list[dict] = []
    for name, spec in sorted(concerns_def.items()):
        script_rel = (spec or {}).get("script") or ""
        script_path = REPO_ROOT / script_rel if script_rel else None
        state_file = _concern_state_file(state_dir, name)
        last_feedback = _latest_feedback(state_dir, name)
        concerns.append({
            "concern": name,
            "fail_closed": bool((spec or {}).get("fail_closed", False)),
            "script": script_rel or None,
            "script_present": bool(script_path and script_path.is_file()),
            "state_file": _rel(state_file, project_root),
            "last_feedback": _rel(last_feedback, project_root),
        })

    trampolines: list[dict] = []
    for row in matrix["platforms"]:
        platform = row["platform"]
        needs_trampoline = bool(row["bindings"]) and platform not in NATIVE_DISPATCH_PLATFORMS
        tpath = _trampoline_for(platform)
        trampolines.append({
            "platform": platform,
            "expected": _rel(tpath, REPO_ROOT),
            "present": tpath.is_file(),
            "required": needs_trampoline,
            "missing": needs_trampoline and not tpath.is_file(),
        })

    # Phase 1 of road-to-hooks-actually-fire-in-consumers: surface
    # the dispatch-issues log so users see hooks that tried and failed.
    state_root = REPO_ROOT / STATE_DIR_DEFAULT
    issues: list[dict] = []
    try:
        sys.path.insert(0, str(REPO_ROOT / "scripts" / "hooks"))
        from dispatch_issues import read_dispatch_issues  # noqa: PLC0415
        issues = read_dispatch_issues(REPO_ROOT)[-20:]  # last 20
    except (ImportError, OSError):
        issues = []

    return {
        "schema_version": 1,
        "platforms": matrix["platforms"],
        "concerns": concerns,
        "trampolines": trampolines,
        "dispatch_issues": issues,
    }


def _render_table(payload: dict) -> str:
    lines: list[str] = []
    # Phase 1 CTA — surfaces at the TOP when issues exist, so a user
    # reading the report can't miss it.
    if payload.get("dispatch_issues"):
        n = len(payload["dispatch_issues"])
        lines.append(
            f"⚠️  Hooks tried to fire but couldn't ({n} entry"
            f"{'ies' if n != 1 else 'y'} in dispatch-issues.jsonl) — "
            "run `./agent-config hooks:install --claude --regen` "
            "(or follow the per-concern hints below)"
        )
        lines.append("")
    lines.append(hooks_status._render_table(payload))
    lines.append("")
    lines.append("Concerns")
    lines.append("-" * 60)
    for c in payload["concerns"]:
        posture = "fail-closed" if c["fail_closed"] else "fail-open"
        script_mark = "✅ " if c["script_present"] else "❌ "
        lines.append(f"{script_mark}{c['concern']:<22} {posture:<11} {c['script'] or '(no script)'}")
        if c["state_file"]:
            lines.append(f"    state:    {c['state_file']}")
        if c["last_feedback"]:
            lines.append(f"    feedback: {c['last_feedback']}")
    lines.append("")
    lines.append("Trampolines")
    lines.append("-" * 60)
    for t in payload["trampolines"]:
        marker = "❌ " if t["missing"] else ("·  " if not t["required"] else "✅ ")
        suffix = "" if t["required"] else "  (not required)"
        lines.append(f"{marker}{t['platform']:<9} {t['expected']}{suffix}")
    # Dispatch-issues detail — last 20 grouped by concern.
    if payload.get("dispatch_issues"):
        lines.append("")
        lines.append("Dispatch issues (last 20)")
        lines.append("-" * 60)
        grouped: dict[str, list[dict]] = {}
        for entry in payload["dispatch_issues"]:
            grouped.setdefault(entry.get("hook", "?"), []).append(entry)
        for hook, entries in sorted(grouped.items()):
            lines.append(f"⚠️  {hook}: {len(entries)} issue(s)")
            # Show the most recent reason + resolution per concern.
            latest = entries[-1]
            lines.append(f"    {latest.get('issue')}: {latest.get('detail')}")
            lines.append(f"    fix → {latest.get('resolution')}")
    return "\n".join(lines)


def _final_exit_code(payload: dict, strict: bool) -> int:
    if not strict:
        return 0
    rc = hooks_status._final_exit_code(payload, strict)
    if rc:
        return rc
    if any(t["missing"] for t in payload["trampolines"]):
        return 1
    if any(not c["script_present"] for c in payload["concerns"]):
        return 1
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--format", choices=["table", "json"], default="table")
    parser.add_argument("--project-root", default=".",
                        help="Project root to inspect (default: cwd)")
    parser.add_argument("--manifest", default=str(dispatch_hook.MANIFEST_PATH))
    parser.add_argument("--strict", action="store_true",
                        help="Exit non-zero on missing bridges, trampolines, "
                             "or concern scripts (CI-friendly).")
    args = parser.parse_args(argv)

    manifest_path = Path(args.manifest)
    if not manifest_path.exists():
        sys.stderr.write(f"hooks_doctor: manifest missing at {manifest_path}\n")
        return 2
    manifest = dispatch_hook._load_yaml(manifest_path)
    project_root = Path(args.project_root).resolve()
    payload = collect(project_root, manifest)

    if args.format == "json":
        print(json.dumps(payload, indent=2, sort_keys=True))
    else:
        print(_render_table(payload))
    return _final_exit_code(payload, args.strict)


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/hooks_doctor.ts
#!/usr/bin/env node
/**
 * Hook doctor — read-only diagnostic over the hook runtime.
 *
 * TypeScript twin of `src/scripts/hooks_doctor.py` (ADR-094 — Python→TS
 * migration, Phase 6 / hooks). Public API mirrors the Python module
 * exactly (snake_case kept deliberately — fidelity over TS idiom).
 *
 * Wraps `scripts/hooks_status.py` (bridge presence + manifest bindings)
 * and adds three diagnostics the bare status table does not surface:
 *
 *   * **Concerns** — every concern declared in the manifest, its
 *     `fail_closed` posture, the on-disk script path, and a one-line
 *     file-exists check.
 *   * **Trampolines** — per-platform shell trampoline expected under
 *     `scripts/hooks/<platform>-dispatcher.sh`; flags any platform that
 *     has manifest bindings but no trampoline on disk.
 *   * **Last feedback** — for each concern, the most-recent dispatcher
 *     feedback file under `agents/runtime/state/.dispatcher/<session>/<concern>.json`,
 *     plus the per-rule state file under `agents/runtime/state/<concern>.json`
 *     when one exists.
 *
 * This is a **read-only** report. It never installs, modifies, or runs
 * anything — same contract as `hooks_status.py`. CI uses `--strict` to
 * turn missing bindings / trampolines into a non-zero exit.
 *
 * Schema: docs/contracts/hook-architecture-v1.md.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

import { read_dispatch_issues } from "./hooks/dispatch_issues.js";
import { type JsonObject, type JsonValue, _load_yaml } from "./hooks/dispatch_hook.js";
import * as hooks_status from "./hooks_status.js";

// src/scripts/hooks_doctor.ts → parents[2] (../../..) is the repo root,
// mirroring the Python `Path(__file__).resolve().parents[2]`.
const REPO_ROOT = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "..",
  "..",
);

// `dispatch_hook.MANIFEST_PATH` is not exported from the TS twin; re-derive
// it the same way (REPO_ROOT/src/scripts/hook_manifest.yaml).
const MANIFEST_PATH = path.join(
  REPO_ROOT,
  "src",
  "scripts",
  "hook_manifest.yaml",
);

// Mutable so tests can point the trampoline lookup at an empty dir
// (mirrors the Python monkeypatch of `hooks_doctor.TRAMPOLINE_DIR`).
export let TRAMPOLINE_DIR = path.join(REPO_ROOT, "src", "scripts", "hooks");

/** Test-only override (the Python test uses monkeypatch.setattr). */
export function _set_trampoline_dir(dir: string): void {
  TRAMPOLINE_DIR = dir;
}

const STATE_DIR_DEFAULT = "agents/runtime/state";

// Platforms whose bridge file (settings.json) invokes the universal
// dispatcher directly — no shell trampoline required. Excluded from the
// "missing trampoline" check.
const NATIVE_DISPATCH_PLATFORMS: ReadonlySet<string> = new Set(["claude"]);

export { hooks_status };

function _isFile(p: string): boolean {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}

function _isDir(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function _isObject(v: unknown): v is JsonObject {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function _trampoline_for(platform: string): string {
  return path.join(TRAMPOLINE_DIR, `${platform}-dispatcher.sh`);
}

function _concern_state_file(state_dir: string, concern: string): string | null {
  const target = path.join(state_dir, `${concern}.json`);
  return _isFile(target) ? target : null;
}

/**
 * Return the most-recent dispatcher feedback file for the concern,
 * walking `agents/runtime/state/.dispatcher/<session>/<concern>.json`.
 */
function _latest_feedback(state_dir: string, concern: string): string | null {
  const dispatcher_dir = path.join(state_dir, ".dispatcher");
  if (!_isDir(dispatcher_dir)) {
    return null;
  }
  const candidates: Array<{ p: string; mtime: number }> = [];
  let sessions: string[];
  try {
    sessions = fs.readdirSync(dispatcher_dir);
  } catch {
    return null;
  }
  for (const sess of sessions) {
    const candidate = path.join(dispatcher_dir, sess, `${concern}.json`);
    try {
      const st = fs.statSync(candidate);
      if (st.isFile()) {
        candidates.push({ p: candidate, mtime: st.mtimeMs });
      }
    } catch {
      /* not a file */
    }
  }
  if (candidates.length === 0) {
    return null;
  }
  candidates.sort((a, b) => b.mtime - a.mtime);
  return candidates[0]!.p;
}

function _rel(p: string | null, root: string): string | null {
  if (p === null) {
    return null;
  }
  const resolvedRoot = path.resolve(root);
  const resolvedP = path.resolve(p);
  const rel = path.relative(resolvedRoot, resolvedP);
  // Python: relative_to raises ValueError when p is not under root →
  // falls back to str(path). Mimic: outside-root paths keep absolute form.
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    return p;
  }
  return rel;
}

export interface ConcernRow {
  concern: string;
  fail_closed: boolean;
  script: string | null;
  script_present: boolean;
  state_file: string | null;
  last_feedback: string | null;
}

export interface TrampolineRow {
  platform: string;
  expected: string | null;
  present: boolean;
  required: boolean;
  missing: boolean;
}

export interface DoctorPayload {
  schema_version: number;
  platforms: hooks_status.PlatformRow[];
  concerns: ConcernRow[];
  trampolines: TrampolineRow[];
  dispatch_issues: JsonObject[];
}

/** Build the doctor payload — JSON-serialisable. */
export function collect(
  project_root: string,
  manifest: JsonObject,
  state_dir_rel: string = STATE_DIR_DEFAULT,
): DoctorPayload {
  const matrix = hooks_status.collect(project_root, manifest);
  const state_dir = path.join(project_root, state_dir_rel);

  const concerns_def = _isObject(manifest["concerns"])
    ? (manifest["concerns"] as JsonObject)
    : {};
  const concerns: ConcernRow[] = [];
  for (const name of Object.keys(concerns_def).sort()) {
    const specRaw = concerns_def[name];
    const spec = _isObject(specRaw) ? specRaw : {};
    const script_rel = typeof spec["script"] === "string" ? spec["script"] : "";
    const script_path = script_rel ? path.join(REPO_ROOT, script_rel) : null;
    const state_file = _concern_state_file(state_dir, name);
    const last_feedback = _latest_feedback(state_dir, name);
    concerns.push({
      concern: name,
      fail_closed: Boolean(spec["fail_closed"] ?? false),
      script: script_rel || null,
      script_present: Boolean(script_path && _isFile(script_path)),
      state_file: _rel(state_file, project_root),
      last_feedback: _rel(last_feedback, project_root),
    });
  }

  const trampolines: TrampolineRow[] = [];
  for (const row of matrix.platforms) {
    const platform = row.platform;
    const needs_trampoline =
      Object.keys(row.bindings).length > 0 &&
      !NATIVE_DISPATCH_PLATFORMS.has(platform);
    const tpath = _trampoline_for(platform);
    const present = _isFile(tpath);
    trampolines.push({
      platform,
      expected: _rel(tpath, REPO_ROOT),
      present,
      required: needs_trampoline,
      missing: needs_trampoline && !present,
    });
  }

  // Phase 1 of road-to-hooks-actually-fire-in-consumers: surface
  // the dispatch-issues log so users see hooks that tried and failed.
  let issues: JsonObject[] = [];
  try {
    issues = read_dispatch_issues(REPO_ROOT).slice(-20) as unknown as JsonObject[];
  } catch {
    issues = [];
  }

  return {
    schema_version: 1,
    platforms: matrix.platforms,
    concerns,
    trampolines,
    dispatch_issues: issues,
  };
}

function _strVal(v: JsonValue | undefined): string {
  return v === undefined || v === null ? "undefined" : String(v);
}

export function _render_table(payload: DoctorPayload): string {
  const lines: string[] = [];
  // Phase 1 CTA — surfaces at the TOP when issues exist, so a user
  // reading the report can't miss it.
  if (payload.dispatch_issues.length > 0) {
    const n = payload.dispatch_issues.length;
    lines.push(
      `⚠️  Hooks tried to fire but couldn't (${n} entr` +
        `${n !== 1 ? "ies" : "y"} in dispatch-issues.jsonl) — ` +
        "run `./agent-config hooks:install --claude --regen` " +
        "(or follow the per-concern hints below)",
    );
    lines.push("");
  }
  lines.push(hooks_status._render_table({
    schema_version: payload.schema_version,
    platforms: payload.platforms,
  }));
  lines.push("");
  lines.push("Concerns");
  lines.push("-".repeat(60));
  for (const c of payload.concerns) {
    const posture = c.fail_closed ? "fail-closed" : "fail-open";
    const script_mark = c.script_present ? "✅ " : "❌ ";
    lines.push(
      `${script_mark}${c.concern.padEnd(22)} ${posture.padEnd(11)} ${
        c.script || "(no script)"
      }`,
    );
    if (c.state_file) {
      lines.push(`    state:    ${c.state_file}`);
    }
    if (c.last_feedback) {
      lines.push(`    feedback: ${c.last_feedback}`);
    }
  }
  lines.push("");
  lines.push("Trampolines");
  lines.push("-".repeat(60));
  for (const t of payload.trampolines) {
    const marker = t.missing ? "❌ " : !t.required ? "·  " : "✅ ";
    const suffix = t.required ? "" : "  (not required)";
    lines.push(`${marker}${t.platform.padEnd(9)} ${t.expected}${suffix}`);
  }
  // Dispatch-issues detail — last 20 grouped by concern.
  if (payload.dispatch_issues.length > 0) {
    lines.push("");
    lines.push("Dispatch issues (last 20)");
    lines.push("-".repeat(60));
    const grouped: Record<string, JsonObject[]> = {};
    for (const entry of payload.dispatch_issues) {
      const hook = _strVal(entry["hook"] ?? "?");
      (grouped[hook] ??= []).push(entry);
    }
    for (const hook of Object.keys(grouped).sort()) {
      const entries = grouped[hook]!;
      lines.push(`⚠️  ${hook}: ${entries.length} issue(s)`);
      // Show the most recent reason + resolution per concern.
      const latest = entries[entries.length - 1]!;
      lines.push(
        `    ${_strVal(latest["issue"])}: ${_strVal(latest["detail"])}`,
      );
      lines.push(`    fix → ${_strVal(latest["resolution"])}`);
    }
  }
  return lines.join("\n");
}

export function _final_exit_code(payload: DoctorPayload, strict: boolean): number {
  if (!strict) {
    return 0;
  }
  const rc = hooks_status._final_exit_code(
    { schema_version: payload.schema_version, platforms: payload.platforms },
    strict,
  );
  if (rc) {
    return rc;
  }
  if (payload.trampolines.some((t) => t.missing)) {
    return 1;
  }
  if (payload.concerns.some((c) => !c.script_present)) {
    return 1;
  }
  return 0;
}

interface ParsedArgs {
  format: "table" | "json";
  project_root: string;
  manifest: string;
  strict: boolean;
}

function parse_args(argv: string[]): ParsedArgs {
  let format: "table" | "json" = "table";
  let project_root = ".";
  let manifest = MANIFEST_PATH;
  let strict = false;
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--format") {
      const v = argv[i + 1];
      if (v === "json" || v === "table") format = v;
      i += 1;
    } else if (arg === "--project-root") {
      project_root = argv[i + 1] ?? ".";
      i += 1;
    } else if (arg === "--manifest") {
      manifest = argv[i + 1] ?? MANIFEST_PATH;
      i += 1;
    } else if (arg === "--strict") {
      strict = true;
    }
  }
  return { format, project_root, manifest, strict };
}

// Python json.dumps(payload, indent=2, sort_keys=True) byte-for-byte.
function _py_json_dumps_sorted(value: JsonValue, indent = 2): string {
  const pad = " ".repeat(indent);
  const esc = (s: string): string => JSON.stringify(s);
  const render = (v: JsonValue, depth: number): string => {
    if (v === null) return "null";
    if (typeof v === "boolean") return v ? "true" : "false";
    if (typeof v === "number") return String(v);
    if (typeof v === "string") return esc(v);
    const curPad = pad.repeat(depth + 1);
    const closePad = pad.repeat(depth);
    if (Array.isArray(v)) {
      if (v.length === 0) return "[]";
      const items = v.map((it) => curPad + render(it, depth + 1));
      return "[\n" + items.join(",\n") + "\n" + closePad + "]";
    }
    const obj = v as JsonObject;
    const keys = Object.keys(obj).sort();
    if (keys.length === 0) return "{}";
    const items = keys.map(
      (k) => curPad + esc(k) + ": " + render(obj[k] as JsonValue, depth + 1),
    );
    return "{\n" + items.join(",\n") + "\n" + closePad + "}";
  };
  return render(value, 0);
}

export function main(argv?: string[]): number {
  const args = parse_args(argv ?? process.argv.slice(2));

  const manifest_path = args.manifest;
  if (!fs.existsSync(manifest_path)) {
    process.stderr.write(`hooks_doctor: manifest missing at ${manifest_path}\n`);
    return 2;
  }
  const manifest = _load_yaml(manifest_path);
  const project_root = path.resolve(args.project_root);
  const payload = collect(project_root, manifest);

  if (args.format === "json") {
    process.stdout.write(
      _py_json_dumps_sorted(payload as unknown as JsonValue) + "\n",
    );
  } else {
    process.stdout.write(_render_table(payload) + "\n");
  }
  return _final_exit_code(payload, args.strict);
}

const isCliEntry =
  process.argv[1] !== undefined &&
  import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (isCliEntry) {
  process.exit(main(process.argv.slice(2)));
}
